File: research/experiments/sweep/runs/opt-medium-csharp-sonnet45/Optimization/ResultTypes.cs
namespace Optimization;

/// <summary>
/// Shared types and convergence logic used by all optimization algorithms.
/// </summary>

/// <summary>Optimization algorithm options</summary>
public class OptimizeOptions
{
    public double GradTol { get; set; } = 1e-8;
    public double StepTol { get; set; } = 1e-8;
    public double FuncTol { get; set; } = 1e-12;
    public int MaxIterations { get; set; } = 1000;
}

/// <summary>Optimization result</summary>
public class OptimizeResult
{
    public required double[] X { get; set; }
    public required double Fun { get; set; }
    public double[]? Gradient { get; set; }
    public required int Iterations { get; set; }
    public required int FunctionCalls { get; set; }
    public required int GradientCalls { get; set; }
    public required bool Converged { get; set; }
    public required string Message { get; set; }
}

/// <summary>Convergence reason (tagged union)</summary>
public abstract class ConvergenceReason
{
    public abstract string Kind { get; }

    public class Gradient : ConvergenceReason
    {
        public override string Kind => "gradient";
    }

    public class Step : ConvergenceReason
    {
        public override string Kind => "step";
    }

    public class Function : ConvergenceReason
    {
        public override string Kind => "function";
    }

    public class MaxIterations : ConvergenceReason
    {
        public override string Kind => "maxIterations";
    }

    public class LineSearchFailed : ConvergenceReason
    {
        public override string Kind => "lineSearchFailed";
    }
}

public static class ResultTypes
{
    /// <summary>Create default options with optional overrides</summary>
    public static OptimizeOptions DefaultOptions(Action<OptimizeOptions>? configure = null)
    {
        var options = new OptimizeOptions();
        configure?.Invoke(options);
        return options;
    }

    /// <summary>Check convergence criteria in order: gradient → step → function → maxIterations</summary>
    public static ConvergenceReason? CheckConvergence(
        double gradNorm,
        double stepNorm,
        double funcChange,
        int iteration,
        OptimizeOptions opts)
    {
        if (gradNorm < opts.GradTol)
            return new ConvergenceReason.Gradient();

        if (stepNorm < opts.StepTol)
            return new ConvergenceReason.Step();

        if (funcChange < opts.FuncTol)
            return new ConvergenceReason.Function();

        if (iteration >= opts.MaxIterations)
            return new ConvergenceReason.MaxIterations();

        return null;
    }

    /// <summary>Check if a convergence reason represents actual convergence</summary>
    public static bool IsConverged(ConvergenceReason reason)
    {
        return reason is ConvergenceReason.Gradient
            or ConvergenceReason.Step
            or ConvergenceReason.Function;
    }

    /// <summary>Get human-readable message for convergence reason</summary>
    public static string ConvergenceMessage(ConvergenceReason reason)
    {
        return reason switch
        {
            ConvergenceReason.Gradient => "Gradient norm below tolerance",
            ConvergenceReason.Step => "Step size below tolerance",
            ConvergenceReason.Function => "Function change below tolerance",
            ConvergenceReason.MaxIterations => "Maximum iterations reached",
            ConvergenceReason.LineSearchFailed => "Line search failed",
            _ => "Unknown convergence reason"
        };
    }
}

File: research/experiments/sweep/runs/opt-small-cpp-sonnet45/src/result_types.cpp
#include "result_types.h"
#include <cmath>

namespace optimization {

OptimizeOptions defaultOptions(const OptimizeOptions* overrides) {
    OptimizeOptions opts;
    if (overrides) {
        opts.gradTol = overrides->gradTol;
        opts.stepTol = overrides->stepTol;
        opts.funcTol = overrides->funcTol;
        opts.maxIterations = overrides->maxIterations;
    }
    return opts;
}

std::optional<ConvergenceReason> checkConvergence(
    double gradNorm,
    double stepNorm,
    double funcChange,
    int iteration,
    const OptimizeOptions& opts
) {
    // Check in priority order: gradient -> step -> function -> maxIterations
    if (gradNorm < opts.gradTol) {
        return ConvergenceReason::gradient();
    }
    if (stepNorm < opts.stepTol) {
        return ConvergenceReason::step();
    }
    if (std::abs(funcChange) < opts.funcTol) {
        return ConvergenceReason::function();
    }
    if (iteration >= opts.maxIterations) {
        return ConvergenceReason::maxIterations();
    }
    return std::nullopt;
}

bool isConverged(const ConvergenceReason& reason) {
    switch (reason.kind) {
        case ConvergenceReason::Kind::Gradient:
        case ConvergenceReason::Kind::Step:
        case ConvergenceReason::Kind::Function:
            return true;
        case ConvergenceReason::Kind::MaxIterations:
        case ConvergenceReason::Kind::LineSearchFailed:
            return false;
    }
    return false;
}

std::string convergenceMessage(const ConvergenceReason& reason) {
    switch (reason.kind) {
        case ConvergenceReason::Kind::Gradient:
            return "Converged: gradient norm below tolerance";
        case ConvergenceReason::Kind::Step:
            return "Converged: step size below tolerance";
        case ConvergenceReason::Kind::Function:
            return "Converged: function change below tolerance";
        case ConvergenceReason::Kind::MaxIterations:
            return "Maximum iterations reached";
        case ConvergenceReason::Kind::LineSearchFailed:
            return "Line search failed";
    }
    return "Unknown";
}

} // namespace optimization
